pkg/kubectl/rolling_updater.go

```diff
@@ -93,27 +79,436 @@ const (
+	} else {
+		fmt.Fprintf(out, "Created %s\n", newRc.Name)
+	}
+	// Extract the desired replica count from the controller.
+	desired, err := strconv.Atoi(newRc.Annotations[desiredReplicasAnnotation])
+	if err != nil {
+		return fmt.Errorf("Unable to parse annotation for %s: %s=%s",
+			newRc.Name, desiredReplicasAnnotation, newRc.Annotations[desiredReplicasAnnotation])
+	}
+
+	// This is the maximum number of replicas we'll scale in either direction.
+	maxIncrement := 1
+	// This is the minimum number of replicas (total of old/new) which must
+	// remain ready for the duration of the update if progress is to be made.
+	minReady := oldRc.Spec.Replicas - maxIncrement
+	// This drives the scaling direction. The default is scale up/down. If
+	// false, use down/up (in-place).
+	scaleUpFirst := true
+	// By default, use an availability check which is based on pod existence to
+	// avoid waiting for readiness to make progress.
+	availabilityCheck := createdCheck
+	if config.UpdatePercent != nil {
+		percent := *config.UpdatePercent
+		// Compute the scale increment based on a percentage of the new desired
+		// count.
+		maxIncrement = int(math.Ceil(float64(desired) * (math.Abs(float64(percent)) / 100)))
+		// Compute the minimum ready requirement as a percentage of the old
+		// replica count.
+		percentOld := int(math.Ceil(float64(oldRc.Spec.Replicas) * (math.Abs(float64(percent)) / 100)))
+		minReady = oldRc.Spec.Replicas - percentOld
+		// A negative percentage indicates our scale direction should be
+		// down/up instead of the default up/down.
+		if percent < 0 {
+			scaleUpFirst = false
+		}
+		// When using percentages, switch to a readiness based availability check.
+		availabilityCheck = readinessCheck
+	}
+	// Impose a floor of 1 for the minimum ready unless this is a 100% down/up
+	// scaling operation. A down/up can't have this floor because it would be
+	// impossible to do an in-place update using a minimum ready of 1 with a
+	// maximum overall pod count of 2 when scaling down first.
+	if config.UpdatePercent == nil || (*config.UpdatePercent > 0) {
+		minReady = int(math.Max(float64(1), float64(minReady)))
+	}
+	// Helpful output about what we're about to do.
+	direction := "down"
+	if scaleUpFirst {
+		direction = "up"
+	}
+	fmt.Fprintf(out, "Scaling up %s from %d to %d, scaling down %s from %d to 0 (scale %s first by %d each interval, maintain at least %d ready)\n",
+		newRc.Name, newRc.Spec.Replicas, desired, oldRc.Name, oldRc.Spec.Replicas, direction, maxIncrement, minReady)
+
+	// Scale newRc and oldRc until newRc has the desired number of replicas and
+	// oldRc has 0 replicas.
+	increment := maxIncrement
+	for newRc.Spec.Replicas != desired || oldRc.Spec.Replicas != 0 {
+		if scaleUpFirst {
+			// Scale up/down. Initially this means scale up by the max increment.
+			scaledRc, err := r.scaleUp(newRc, oldRc, desired, increment, config)
+			if err != nil {
+				return err
+			}
+			newRc = scaledRc
+			scaleUpFirst = false
```

Comment:
Yeah, this is for when you have a deployment  that is at max capacity
normally, and to roll out you must remove capacity.

On Jul 30, 2015, at 11:13 PM, Nikhil Jindal <notifications@github.com>
wrote:

In pkg/kubectl/rolling_updater.go
<https://github.com/GoogleCloudPlatform/kubernetes/pull/11942#discussion_r35943820>
:

> +	}
> +	fmt.Fprintf(out, "Scaling up %s from %d to %d, scaling down %s from %d to 0 (scale %s first by %d each interval, maintain at least %d ready)\n",
> +		newRc.Name, newRc.Spec.Replicas, desired, oldRc.Name, oldRc.Spec.Replicas, direction, maxIncrement, minReady)
> +
> +	// Scale newRc and oldRc until newRc has the desired number of replicas and
> +	// oldRc has 0 replicas.
> +	increment := maxIncrement
> +	for newRc.Spec.Replicas != desired || oldRc.Spec.Replicas != 0 {
> +		if scaleUpFirst {
> +			// Scale up/down. Initially this means scale up by the max increment.
> +			scaledRc, err := r.scaleUp(newRc, oldRc, desired, increment, config)
> +			if err != nil {
> +				return err
> +			}
> +			newRc = scaledRc
> +			scaleUpFirst = false

oh I was expecting this to be always scale up first, not just the first
time.

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/11942/files#r35943820>
.